core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -811,18 +795,43 @@ object IO {
         } yield fiberA.zipWith(fiberAs)(_ :: _)
     }
 
+  /**
+   * Acquires a resource, do some work with it, and then release that resource. With `bracket0`
+   * not only is the acquired resource be cleaned up, the outcome of the computation is also
+   * reified for processing.
+   */
+  final def bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, ExitResult[E, B]) => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
+    IORef[E, Option[(A, ExitResult[E, B])]](None).flatMap { m =>
+      (for {
+        a <- acquire
+              .flatMap(a => m.write[E](Some((a, ExitResult.Terminated(new Throwable("Interrupted fiber"))))).const(a))
+              .uninterruptibly
+        b <- use(a).run.flatMap(
+              r =>
```

Comment:
I thought of a quick fix for this problem: change `release : (A, Option[Either[E, B]]) => Infallible[Unit]`.

Then instead of `use(a).run`, do `use(a).attempt`.

We aren't promising to return the `Throwable`, but then again, the `Throwable` we return is not useful. I think this is the best we can do for now and is an improvement since it makes `use(a)` interruptible.

Reply:
Two birds with one stone 💯